### DISCUSSION TITLE
i updated streamlink on my raspberry pi and now it does not work anymore

### DISCUSSION THREAD

**Discussion by zerger3** [streamlink/streamlink#3406] [Q&A] [answered]
## Bug Report

- [ x] This is a bug report and I have read the contribution guidelines.
- [ x] I am using the latest development version from the master branch.

### Description

i think the issue is that the old version worked with python 2 and the new with 3

### Expected / Actual behavior

after checking the debian package i noticed its a old version so i uninstalled via apt
i installed first with
sudo pip install --upgrade streamlink
which didnt work and i noticed its python 2, so i uninstalled it with pip and then used 

sudo pip3 install --upgrade streamlink
which was succesful
`Successfully installed isodate-0.6.0 streamlink-1.7.0 websocket-client-0.57.0
`
now when i try to start streamlink i get the error

Traceback (most recent call last):
  File "/home/pi/.local/bin/streamlink", line 6, in <module>
    from streamlink_cli.main import main
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink_cli/main.py", line 21, in <module>
    from streamlink import __version__ as streamlink_version
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink/__init__.py", line 21, in <module>
    from .api import streams
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink/api.py", line 1, in <module>
    from .session import Streamlink
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink/session.py", line 12, in <module>
    from streamlink.logger import StreamlinkLogger, Logger
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink/logger.py", line 7, in <module>
    from streamlink.compat import is_py2
  File "/home/pi/.local/lib/python2.7/site-packages/streamlink/compat.py", line 48, in <module>
    from backports.shutil_which import which
ImportError: No module named shutil_which

how can i fix that?

"

### Additional comments, etc.

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

**Comment by bastimeyer** [comment, marked as answer]
> sudo pip install --upgrade streamlink
> sudo pip3 install --upgrade streamlink

You should **never** install python packages globally with pip, as it'll introduce conflicts with your native package manager at some point. You're supposed to use the `--user` flag, which writes into `$HOME/.local`, or use a virtual environment.

This is explicitly pointed out in the docs:
https://streamlink.github.io/latest/install.html#pypi-package-and-source-code

> sudo pip3 install --upgrade streamlink
> which was succesful

Since you've installed globally, the entry script will be found in `/usr/bin` (or `/usr/local/bin`), not in `$HOME/.local/bin`.

> /home/pi/.local/bin/streamlink
> /home/pi/.local/lib/python2.7/site-packages/streamlink_cli/main.py

You have **locally installed** a different streamlink version based on python 2.7, and `$HOME/.local/bin` is set before `/usr/bin` in your `PATH` environment variable, which means by running `streamlink`, you command line shell will resolve the incorrect path first.

Take a look at the `PATH` env var value:
```
echo $PATH
```

and clean up / uninstall the old local streamlink python27 install
```
pip --user uninstall streamlink
```

**Comment by zerger3** [comment]
thank you for your reply. i only understood 75% but:

pip uninstall  streamlink
pip3 uninstall  streamlink
pip3 install --upgrade --user streamlink
and adding
export PATH="${HOME}/.local/bin:${PATH}" at the end of ~/.profile 
fixed it.
could you please also help me with the config file?
```
it will look in one or both directories
$XDG_CONFIG_HOME/streamlink/config
~/.streamlinkrc
```

do i then create a "config" file in
/home/pi/streamlink ?

**Comment by bastimeyer** [comment]
https://streamlink.github.io/latest/cli.html#configuration-file

> Platform          | Location
> --                | --
> Unix-like (POSIX) | $XDG_CONFIG_HOME/streamlink/config
>                   |  ~/.streamlinkrc
> Windows           | %APPDATA%\streamlink\streamlinkrc

> Note
> `$XDG_CONFIG_HOME` is `~/.config` if it has not been overridden

Streamlink's main config file is located at `${XDG_CONFIG_HOME:-$HOME/.config}/streamlink/config`, which means `$HOME/.config/streamlink/config` on most systems, as `$XDG_CONFIG_HOME` is empty by default. See the ["XDG Base Directory Specification"](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) ([arch wiki](https://wiki.archlinux.org/index.php/XDG_Base_Directory)) if you have questions about this variable.
`$HOME/.streamlinkrc` is an old and deprecated fallback path.

**Reply by gravyboat** [reply]
@bastimeyer I've moved this over to the discussions (it should have been one to start with) so we can get an idea of what they look like.

**Comment by zerger3** [comment]
thank you for your time. i got it to work now and it seems i managed to create  the config file
the installation process without a simple apt package was very confusing for me. maybe it is possible to improve the installation instructions so linux novices like me have it easier.
thx for creating such a great tool

**Reply by gravyboat** [reply]
> the installation process without a simple apt package was very confusing for me. maybe it is possible to improve the installation instructions so linux novices like me have it easier.

@bastimeyer I talked to someone else recently who is experienced with Linux and while they didn't struggle they also had a similar issue because the apt package is way out of date. Maybe we should stop recommending people use the apt package since it isn't being maintained and suggest a `--user` install of Streamlink instead?

**Reply by zerger3** [reply]
maybe also add a "how to update from an old version" and "how to check if installation method x will install the current version"

**Reply by bastimeyer** [reply]
@gravyboat That's the problem with point-release distros like Debian and Ubuntu. Packages won't be upgraded for older releases and distros which are built on top of these distros require additional time until the newer package versions are available to their users. And then there's also the issue with custom package repos or PPAs which have to be added by the user. That's why I don't recommend using these distros on desktop systems. It is not ideal for endusers or software developers.

- https://packages.debian.org/search?keywords=streamlink
- https://packages.ubuntu.com/search?keywords=streamlink
- https://launchpad.net/~nilarimogard/+archive/ubuntu/webupd8/+packages?field.name_filter=streamlink

Recommending installs via pip also has its problems, especially for non-tech-savvy people, as it requires additional setup and also comes with maintenance burden. And there's still the distinction between py2 and py3 on these distros, which can be confusing for some people, as you can see in the OP.

----

@zerger3 

> maybe also add a "how to update from an old version"

This is not necessary. Your problem were two separate installs which you were not aware of. Upgrading via your native package manager is trivial (if a newer package version is available), and via pip, there's the `--upgrade` / `-U` parameter, which we have already listed several times in the docs. The Streamlink docs can't cover every edge case on every OS that comes with Python package installing.

> "how to check if installation method x will install the current version"

This is already implied in the first paragraph of the PyPi package section. It could be improved though.

**Reply by gravyboat** [reply]
@bastimeyer Oh yeah I'm well aware of the issues with older releases of certain distros. You're right though that suggesting an install via pip may be too difficult for some users.

**Comment by WolfganP** [comment]
Following on this topic with a similar question. Is it possible to install the stable version (via pip) side by side to a development version from a cloned repository?

I tried with repository cloning and then `python3 setup.py install --user` as per https://github.com/streamlink/streamlink/issues/2896#issuecomment-620595417 (besides discussion at https://github.com/streamlink/streamlink/issues/3020 and the official docs https://streamlink.github.io/latest/install.html#pypi-package-and-source-code) but I can't avoid mangling both installs even modding https://github.com/streamlink/streamlink/blob/master/setup.py

I was able to do it with youtube-dl as a "portable" version (zip packed) with a diff name but that didn't work with streamlink.
Any tip?

**Reply by bastimeyer** [reply]
> Is it possible to install the stable version (via pip) side by side to a development version

Set up a virtual environment:
https://streamlink.github.io/latest/install.html#virtual-environment

And install it from any source you want:
https://streamlink.github.io/latest/install.html#pypi-package-and-source-code

**Reply by WolfganP** [reply]
Thx @bastimeyer I tried the virtual env path and it works but is extremely slow in a Raspberry, that's why I was looking for a way to simply have a diff app version in the path (ie streamlinkdev) at the same time than the official release.

**Reply by bastimeyer** [reply]
Why would it be extremely slow in comparison to the regular system-wide environment?

Not sure if this will work, but you could try changing the `PYTHONPATH` env var to a custom/secondary virtual-env path and then run streamlink via `python -s -m streamlink` by using the system python executable, assuming your custom module path was built with the same python env version, of course.

```bash
env PYTHONPATH=/path/to/virtual-env/lib/python3.9/site-packages/ python -s -m streamlink --version
```

- https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH
- https://docs.python.org/3/using/cmdline.html#cmdoption-s

btw, this is all very much off topic of this thread and doesn't belong here...

**Reply by WolfganP** [reply]
I'll try that to optimize the alt env. Thanks!
